### PR TITLE
Add 20.2.5

### PR DESCRIPTION
**Pull request by altendky**

### CODE REVIEW

**Comment by altendky** [comment]
Kick CI for a published build.

**Comment by altendky** [comment]
below are links to the branch off development to test and a build there, and the branch of dave's to test and a build there.  aside from the builds passing, i have looked and confirmed that the branch off dave's is using 20.2.5 command lines and not 18.12.3.  also, the branch off development is the opposite.

https://github.com/epcpower/grid-tied/commits/feature/docker-ccstudio-pr12
https://github.com/epcpower/grid-tied/runs/2720152090
https://github.com/epcpower/grid-tied/commits/feature/threeLevelMixedModulation-test-20.2.5
https://github.com/epcpower/grid-tied/runs/2679925184

File: docker.py [hunk 1]
```diff
@@ -42,75 +75,141 @@ def main(tarball, install_iu, uninstall_iu):
...
-        print('No IU specified for installation')
-
-    iu_to_uninstall = uninstall_iu
-    if len(iu_to_uninstall) > 0:
-        print('Uninstalling IU {}'.format(iu_to_uninstall))
-        uninstall_iu(iu=iu_to_uninstall, ccstudio=ccstudio)
-    else:
-        print('No IU specified for uninstallation')
+    with virtual_display():
+        print_flush('+++ here')
```
Comment (epcAnker) [inline docker.py]: Can take this debug print out?

File: docker.py [hunk 2]
```diff
@@ -123,16 +222,26 @@ def process_iu(iu, ccstudio, command):
         '-display', ':0',
     ]
 
+    xvfb_process = subprocess.Popen(xvfb_command)
+    print_flush('+++ somewhere')
```
Comment (epcAnker) [inline docker.py]: remove debug

File: docker.py [hunk 3]
```diff
@@ -123,16 +222,26 @@ def process_iu(iu, ccstudio, command):
...
+    try:
+        x11vnc_process = subprocess.Popen(x11vnc_command)
+        print_flush('+++ there')
```
Comment (epcAnker) [inline docker.py]: remove debug

File: docker.py [hunk 4]
```diff
@@ -42,75 +75,141 @@ def main(tarball, install_iu, uninstall_iu):
...
+                ],
+                check=True
+            )
+            print_flush('success--------------------------------')
+        except subprocess.CalledProcessError:
+            install_logs = installed/'install_logs'
+            for parent in install_logs.parents:
+                print_flush('checking: ', parent, parent.exists())
+            install_logs = list(
+                path
```
Comment (epcAnker) [inline docker.py]: formatting, should this all be on one line?

File: docker.py [hunk 5]
```diff
@@ -42,75 +75,141 @@ def main(tarball, install_iu, uninstall_iu):
...
+                for path in install_logs.rglob('*')
+                if path.is_file()
+            )
+            print_flush('potential install logs:')
+            for maybe in install_logs:
```
Comment (epcAnker) [inline docker.py]: I think this for statement can be removed

File: docker.py [hunk 6]
```diff
@@ -42,75 +75,141 @@ def main(tarball, install_iu, uninstall_iu):
...
+            print_flush(list(parent.iterdir()))
+
+        link = pathlib.Path(os.sep)/'usr'/'local'/'bin'/'ccstudio'
+        link.symlink_to(ccstudio)
+
+        installed_ius = get_installed_ius(ccstudio=ccstudio)
+        ius_to_install = requested_install_ius - installed_ius
+
+        needed_uninstall_ius = {
+            installed_iu
```
Comment (epcAnker) [inline docker.py:134]: Can you add a comment of what this is doing?  Hard to read & understand

**Review by epcAnker — APPROVED** [review]
view comments

**Comment by altendky** [comment]
Yeah, I was being a bit lax here...  thanks for calling me out.  I'll tidy it up.

File: docker.py (hunk 4 above)
Comment (altendky) [reply docker.py]: i'll read that as "start using black!".  `:]`  (and change it for now)  (but no, black wouldn't put it on one line, though it also wouldn't format it per my preference which is what is there)

File: docker.py (hunk 6 above)
Comment (altendky) [reply docker.py:134]: Yep, that comprehension got a bit out of hand.

**Comment by altendky** [comment]
I reran the linked test builds and they are acting as before.  Both builds passed as well.

**Pull request merged**